File: sfas-ui/src/app/core/services/vendor.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Vendor } from '../models/vendor.model';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class VendorService {
  private http = inject(HttpClient);
  private baseUrl = environment.apiUrl + '/api/v1/vendors';

  getAllVendors(): Observable<Vendor[]> {
    return this.http.get<Vendor[]>(this.baseUrl);
  }

  getVendorById(id: string): Observable<Vendor> {
    return this.http.get<Vendor>(`${this.baseUrl}/${id}`);
  }

  createVendor(vendor: Partial<Vendor>): Observable<Vendor> {
    return this.http.post<Vendor>(this.baseUrl, vendor);
  }

  updateVendor(id: string, vendor: Partial<Vendor>): Observable<Vendor> {
    return this.http.put<Vendor>(`${this.baseUrl}/${id}`, vendor);
  }

  deleteVendor(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: sfas_mobile/lib/features/vendors/services/vendor_service.dart
import 'package:dio/dio.dart';
import '../../../core/constants/endpoints.dart';
import '../models/vendor.dart';

class VendorService {
  final Dio _dio;

  VendorService(this._dio);

  Future<List<Vendor>> getVendors() async {
    try {
      final response = await _dio.get(Endpoints.vendors);
      final List<dynamic> data = response.data;
      return data.map((json) => Vendor.fromJson(json)).toList();
    } catch (e) {
      rethrow;
    }
  }

  Future<Vendor> createVendor(Vendor vendor) async {
    try {
      final response = await _dio.post(
        Endpoints.vendors,
        data: vendor.toJson(),
      );
      return Vendor.fromJson(response.data);
    } catch (e) {
      rethrow;
    }
  }

  Future<Vendor> updateVendor(String id, Vendor vendor) async {
    try {
      final response = await _dio.put(
        Endpoints.vendorDetails(id),
        data: vendor.toJson(),
      );
      return Vendor.fromJson(response.data);
    } catch (e) {
      rethrow;
    }
  }

  Future<void> deleteVendor(String id) async {
    try {
      await _dio.delete(Endpoints.vendorDetails(id));
    } catch (e) {
      rethrow;
    }
  }
}
